harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -426,17 +429,19 @@ def _train_for_step(
 
             batch_dur = time.time() - batch_start_time
             samples_per_second = batch_inputs / batch_dur
-            if self.hvd_config.use:
-                samples_per_second *= hvd.size()
+            samples_per_second *= self.context.distributed.get_size()
             self.prof.record_metric("samples_per_second", samples_per_second)
             per_batch_metrics.append(tr_metrics)
 
         # Aggregate and reduce training metrics from all the training processes.
-        if self.hvd_config.use and self.hvd_config.average_training_metrics:
+        if (
+            self.context.distributed.size > 1
+            and self.use_horovod
```

Comment:
Same thing about trying to distinguish horovod dtrain from non-horovod dtrain.  As evidence of my claim that "trying to predict the horovod codepaths and the non-horovod codepaths before you actually add a second dtrain backend will mostly cause bugs", you have not been consistent in these checks even in this PR.  For example, context.wrap_optimizer uses horovod but does not have protection.  Having some checks now is more of a false sense of security, IMO.

However, since you are actively working on non-horovod dtrain for pytorch trial, it's really up to you if you want to get rid of these horovod-specific checks with pytorch trial.  You do have enough background to make correct decisions here, which will be meaningful in the near future.  You can take my suggestions here or not.  With TFKeras or Estimator though, I would definitely like you to remove the horovod-specific checks and just have assertions that only the horovod backend is in use.  I really doubt that we ever need to revisit those trial classes, but if we do, it would be better to start with a blank slate, so to speak.

Reply:
since I want to get this change out before introducing native dtrain for pytorch, I went with your suggestion of removing horovod checks and asserting horovod must be used if distributed training.